applicationset/utils/utils.go

```diff
@@ -1,29 +1,159 @@
 package utils
 
 import (
+	"bytes"
 	"encoding/json"
 	"fmt"
 	"io"
 	"reflect"
+	"regexp"
 	"sort"
-	"strconv"
 	"strings"
+	"unsafe"
+
+	"text/template"
 
-	log "github.com/sirupsen/logrus"
 	"github.com/valyala/fasttemplate"
 
+	log "github.com/sirupsen/logrus"
+
 	argoappsv1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
 	argoappsetv1 "github.com/argoproj/argo-cd/v2/pkg/apis/applicationset/v1alpha1"
 )
 
 type Renderer interface {
-	RenderTemplateParams(tmpl *argoappsv1.Application, syncPolicy *argoappsetv1.ApplicationSetSyncPolicy, params map[string]string) (*argoappsv1.Application, error)
+	RenderTemplateParams(tmpl *argoappsv1.Application, syncPolicy *argoappsetv1.ApplicationSetSyncPolicy, params map[string]interface{}, useGoTemplate bool) (*argoappsv1.Application, error)
 }
 
 type Render struct {
 }
 
-func (r *Render) RenderTemplateParams(tmpl *argoappsv1.Application, syncPolicy *argoappsetv1.ApplicationSetSyncPolicy, params map[string]string) (*argoappsv1.Application, error) {
+func copyValueIntoUnexported(destination, value reflect.Value) {
+	reflect.NewAt(destination.Type(), unsafe.Pointer(destination.UnsafeAddr())).
+		Elem().
+		Set(value)
+}
+
+func copyUnexported(copy, original reflect.Value) {
+	var unexported = reflect.NewAt(original.Type(), unsafe.Pointer(original.UnsafeAddr())).Elem()
+	copyValueIntoUnexported(copy, unexported)
+}
+
+// This function is in charge of searching all String fields of the object recursively and apply templating
+// thanks to https://gist.github.com/randallmlough/1fd78ec8a1034916ca52281e3b886dc7
+func (r *Render) deeplyReplace(copy, original reflect.Value, replaceMap map[string]interface{}, useGoTemplate bool) error {
```

Comment:
Could you wrap errors returned by this function?

Reply:
Not sure about wrapping error here. Maybe simply an error message more descriptive in the template part would do the job (and then the user will have to check the field which has its template).

Example:

```
failed to execute go template {{.data.test}}: template: :1:7: executing "" at <.data.test>: can't evaluate field test in type interface {}
```

User will just have to search field with "{{.data.test}}"

Reply:
Done in commit 95bc2b7c5aad9651a532c555342c45ec104ff6de